### ISSUE TITLE
Define Branching Strategy for Spamgourmet. 

### ISSUE THREAD

**Issue by mikedlr** [spamgourmet/spamgourmet#45]
I'd like us to have a simple but explicit branching strategy and change some of the way we handle things. I'm more or less about to do this because the current state of PRs requires it but would like to make sure there's explicit discussion and understanding before I do it and also make it clear that what I'm doing is subject to change if that turns out to be preferred. 

I'd propose

* create a "development" branch - for now that should be the target for PRs - later things should be merged there after basic CI testing has completed
* name a "release" branch - this should point to the latest release which is the thing that people should deploy - it should follow after the "development" branch has been confirmed for release - likely due to automated testing - for now because we say so. 
* other branches will be created as needed by CI processes and so on - to be documented in the docs directory.

N.B. I'd explicitly propose we don't use [GitFlow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow) (please read article) and aim for something close to a trunk based workflow.

**Comment by jgonggrijp** [comment]
What you're describing is already very close to GitFlow, and I would be fine with using that (I use GitFlow for all projects that I didn't inherit with some other branching model already in place). However, I'm fine with any model that separates between an integration branch and a production branch, especially if the release tags are on the latter. It's all roughly equivalent, in my opinion.

The distinguishing feature of GitFlow, compared to what you're describing, is that a `release/x.y.z` branch in split off from the integration branch, in order to bump the version number and dot the i's and cross the t's, before merging into the production branch. I personally find that a sensible way to do things, but I'm also fine with other approaches. Exact names and prefixes for branches are just details, in my opinion.

**Comment by fredericmora** [comment]
Do we even need a release branch? We could be fine with just master used as dev branch.

The workflow used in our company is a single master branch. Devs cut feature branches if needed and merge to master. It works with projects much larger than sg.

That said, if you have reasons to believe that a release branch makes things easier, let me know.

**Comment by jgonggrijp** [comment]
If you only have a single `master` (or `main`) branch that you use for everything, you need to either be very strict with what you allow to be merged into it, or accept that some releases will be unexpectedly buggy, because that branch will be a moving target. Separating integration and production allows you to not worry whether feature branches are perfect, while still keeping tight control over what ends up in the next release. I think #39 and #43 motivate why @mikedlr would like to have that separation.

**Comment by mikedlr** [comment]
So some comments /explanations/responses;
1) what @jgonggrijp said - I'd like to have a clear separation between a branch you can mostly trust (`tested`) and a branch that developers are working on (`development`) 
2) in my vision the branch `development` would match exactly with what @fredericmora describes - everyone would merge to development - all other branches (`tested` / `release`) would be created automatically by non human processes e.g. after automated testing or as the code is being pushed up to spamgourmet.com 
3) ignoring any worries about the term itself the name `master` is a bit ambiguous. Is it the latest place that developers push to or is it the fully tested branch that people should work from? Different projects with different ways of working use it in each meaning. I'd like to get a clearer name for the branches we actually use.
4) in contrast to GitFlow I'd propose that a "release" - as in a set of code that has probably actually been bundled up into a tarball for download and that has been put onto spamgourmet.com would be a tag. The `release` branch would just be an updatable pointer to the latest release for people who want to debug exactly what is in production - the creation of the tag would be fully automated.
5) the last point means that the release rituals of gitflow aren't needed - someone might create a release branch later if someone wants to make specific fixes to problems found on the release, and they can always do that using the tag as a starting point but normally they would just work directly on the `development` branch as @fredericmora described.

**Comment by mikedlr** [comment]
6) developer feature branches - typically prefixed with their initials or some description of a group can do whatever they want with no interference from this process - so if someone is trying to do a release for an urgent fix and you have some code that would disturb that then you just keep that on a "personal" branch until the fix is released.

**Comment by mikedlr** [comment]
7. definitely open for further comments on naming - maybe "production" is a better name for the branch that points to what is currently on spamgourmet.com and there should only be release tags?

**Comment by jgonggrijp** [comment]
@mikedlr Sorry for entering a discussion with you here, but my impression is that you are only falling over the way Vincent Driessen has chosen to name the branches, while the abstract formalism is actually exactly what you want.

> 4\. in contrast to GitFlow I'd propose that a "release" - as in a set of code that has probably actually been bundled up into a tarball for download and that has been put onto spamgourmet.com would be a tag. The `release` branch would just be an updatable pointer to the latest release for people who want to debug exactly what is in production - the creation of the tag would be fully automated.

As I wrote before, this is not a contrast; you're just naming things differently.

> 5\. the last point means that the release rituals of gitflow aren't needed - someone might create a release branch later if someone wants to make specific fixes to problems found on the release, and they can always do that using the tag as a starting point but normally they would just work directly on the `development` branch as @fredericmora described.

Again, there is only a naming difference here. GitFlow would call this a hotfix branch, but regardless of how you name it, it is something you can (and should) do in some situations.

> 6\. developer feature branches - typically prefixed with their initials or some description of a group can do whatever they want with no interference from this process

This is exactly what GitFlow calls a feature branch, just named differently.

> - so if someone is trying to do a release for an urgent fix and you have some code that would disturb that then you just keep that on a "personal" branch until the fix is released.

Here you are turning things around. We have a separation between an integration branch (`development` as you call it) and a stable release branch (`release` or `production` or whatever). There is no need to postpone merging anywhere in the process. The person working on that urgent fix can branch off the integration branch (or if it is urgent, perhaps more likely off the stable release branch). In the meanwhile, it does not matter what happens with the "disturbing" code; it can be freely merged into the integration branch. The fixing branch eventually ends up merged into the stable release branch *and* back into the integration branch, so that all future work takes it into account.

That fixing branch is exactly what GitFlow means by a release branch or a hotfix branch, depending on which branch it started from. We don't have to *name* it `release/x.y.z` or `hotfix/x.y.z`, but the principle to prepare a release on a branch that splits off either of those main branches is just a smart thing to do.

**Comment by mikedlr** [comment]
@jgonggrijp 

> @mikedlr Sorry for entering a discussion with you here,

that's absolutely fine - it's the whole point of starting this discussion and it's really appreciated.

I think the key thing to say as far my proposed difference from normal GitFlow goes is that the `master` branch will not exist at all any more. The development branch (`devel`  maybe?) will replace it completely as a merge point. That's a purely textual / name change and is separate from the philosophy changes proposed which mean that the release and feature **branches** will never normally be created. 

As far as explicit use of GitFlow goes, two articles which say things better than I could

* [Please stop recommending Git Flow!](https://georgestocker.com/2020/03/04/please-stop-recommending-git-flow/)
* [GitFlow considered harmful](https://www.endoflineblog.com/gitflow-considered-harmful) 

please read through and then maybe we should get on a call to discuss?

**Comment by mikedlr** [comment]
Also worth a read - 

* [The problem with Git flow / GitLab Flow: a simplified branching strategy](https://about.gitlab.com/blog/2020/03/05/what-is-gitlab-flow/)

**Comment by jgonggrijp** [comment]
Alright, I read all three articles you linked to. I will concede three points that speak against git-flow in some situations:

- It is sophisticated, and many people will hate anything that they perceive as complex (this appears to be a strong motivation for the authors of the first two articles).
- It has no added value if you want to do continuous delivery (this is something everyone will acknowledge, even Vincent Driessen himself).
- The only added value of the second permanent branch with the stable releases, is that you have a name that always points to the latest stable release (this also appears to be your reason for wanting to have one). I think the second article made a convincing case that you can do something that is very close to git-flow, without having this second permanent branch.

But let's turn this around. I don't think we should be discussing whether git-flow is the right branching model or not. Rather, we should be discussing how we want to organize releases and branches, as an open question with a free-form answer. (So far, I've only been arguing that what you want seems extremely similar to git-flow, and you still haven't convinced me otherwise. But I digress.) I think some points of consensus have already been reached here and on Signal:

- While this has been implicit so far, I don't think anyone finds Spamgourmet a suitable project for continuous delivery. The distinction between releases and non-releases is mostly due to this fact: we do not trust every (merge) commit on the main branch to be safe for release.
- You and I have agreed that a second permanent branch with only the stable releases seems useful.
- Features branches should be rebased on the main branch and then merged with `--no-ff`, to create "empty" merge commits that preserve information and keep the history easy to follow at the same time.

What I don't quite understand yet, is how you want to move from the unsafe main branch to a safe release. Suppose that the main branch contains some new features that you want to release, but there are also some new bugs that make the software potentially insecure (or unusable). Obviously, you will start a new branch to fix those issues and polish the thing into a trustworthy release. While you work on that, somebody else completes another branch, which will introduce a new feature but potentially also new bugs into the main branch. How do you want to proceed in that situation?

**Comment by mikedlr** [comment]
> Rather, we should be discussing how we want to organize releases and branches, as an open question with a free-form answer. 

This. 100%.. 

> I don't think anyone finds Spamgourmet a suitable project for continuous delivery.

I'd split 
 "continuous delivery" - we automatically make a version and put it somewhere including possibly test deploys 
from
"continuous deployment" - we deliver direct to production 

If we talk about continuous deployment, that's definitely true right now. We simply don't have the tests and checks written which would ensure that things kept working when delivered. It's not impossible later but that would be much later and with quite a bit of work done. We should work towards that goal because the closer we get the safer, but that doesn't mean we'll reach it any time soon. 

> second permanent branch with only the stable releases seems useful.

that's fine and I think we agree, but there's a detail which occasionally becomes important  when there's only one branch

branch -  movable to tag to **one** version and implicitly the history of that version but without explicit record of where that branch pointed in the past
tag - fixed pointer to one version 

I think we should tag, at some point, things that are being released. possibly like `release-<commit-hash>`.  If there's a single branch, it's not clear, that, for example, after a rebase or a revert the history of a branch will include a preceding release. Having the tags allows us to find older releases. We might do more clever tagging. 

> Features branches should be rebased on the main branch and then merged with --no-ff, to create "empty" merge commits that preserve information and keep the history easy to follow at the same time.

That's what we agreed on in signal discussion. Sounds like it should be fine for us. 

> What I don't quite understand yet, is how you want to move from the unsafe main branch to a safe release

I think that we'd go with the process that @fredericmora described using where all fixes happen on the main branch. For the period prior to the release we might hold off on merges that might reduce the stability of the main branch and only do smaller bug-fix fixes. 

What I don't think we should have to do is _"polish the thing into a trustworthy release"_. We should try to work on the idea that it's **always** more or less release ready and that we build enough tests and checks that we don't have a problem releasing from the tested branch at more or less any time.

**Comment by jgonggrijp** [comment]
> I'd split "continuous delivery" - we automatically make a version and put it somewhere including possibly test deploys from "continuous deployment" - we deliver direct to production
> 
> If we talk about continuous deployment, that's definitely true right now. We simply don't have the tests and checks written which would ensure that things kept working when delivered. It's not impossible later but that would be much later and with quite a bit of work done. We should work towards that goal because the closer we get the safer, but that doesn't mean we'll reach it any time soon.

You seem to mean to imply that continuous delivery is an option, despite continuous deployment being out of reach at this time. Do you want to issue releases that aren't safe to deploy? If so, how will the main Spamgourmet service distinguish between safe and unsafe releases?

> I think we should tag, at some point, things that are being released.

Yes, of course. It didn't even cross my mind that we might *not* do that! However, now that you have made this explicit, I realize there are no tags in the repository yet.

> possibly like `release-<commit-hash>`.

Commit hashes are very meaningless. Why not just a version number? Semver?

> If there's a single branch, it's not clear, that, for example, after a rebase or a revert the history of a branch will include a preceding release.

Yes that is an additional reason to do tagging, although I don't think we should rebase or revert anything on a branch dedicated to stable releases.

> What I don't think we should have to do is _"polish the thing into a trustworthy release"_. We should try to work on the idea that it's **always** more or less release ready and that we build enough tests and checks that we don't have a problem releasing from the tested branch at more or less any time.

Here, again, you seem to believe that a release doesn't need to be safe. Or if you do, you want to set a very high standard for merges into the main branch.

**Comment by mikedlr** [comment]
> Do you want to issue releases that aren't safe to deploy? If so, how will the main Spamgourmet service distinguish between safe and unsafe releases?

If there is a manual verification step of some sort between the code on GitHub and the deployment then yes, I'd believe that "unsafe" ("less safe"?) releases would be the right thing. They could happily be tagged rc (for release candidate) or beta or something to show that they aren't yet fully tested. Once the verification is complete or at the moment of deploy to production then a better label could be given to show that they are safe(er) to us.

> Here, again, you seem to believe that a release doesn't need to be safe. Or if you do, you want to set a very high standard for merges into the main branch.

I personally worry about reliably having the time to do much safety improvement via manual testing. The automated tests and verification should be set up to ensure a reasonably "safe" release in which case the standard of code going to the main branch should be pretty high.
